internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go

```diff
@@ -0,0 +1,103 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package http
+
+import (
+	"context"
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/atomic"
+)
+
+type downloadProgressReporter struct {
+	sourceURI   string
+	interval    time.Duration
+	warnTimeout time.Duration
+	length      float64
+
+	downloaded atomic.Int
+	started    time.Time
+
+	progressObservers []progressObserver
+}
+
+func newDownloadProgressReporter(sourceURI string, timeout time.Duration, length int, progressObservers ...progressObserver) *downloadProgressReporter {
+	interval := time.Duration(float64(timeout) * downloadProgressIntervalPercentage)
+	if interval == 0 {
+		interval = downloadProgressMinInterval
+	}
+
+	return &downloadProgressReporter{
+		sourceURI:         sourceURI,
+		interval:          interval,
+		warnTimeout:       time.Duration(float64(timeout) * warningProgressIntervalPercentage),
+		length:            float64(length),
+		progressObservers: progressObservers,
+	}
+}
+
+func (dp *downloadProgressReporter) Write(b []byte) (int, error) {
+	n := len(b)
+	dp.downloaded.Add(n)
+	return n, nil
+}
+
+func (dp *downloadProgressReporter) Report(ctx context.Context) {
+	started := time.Now()
+	dp.started = started
+	sourceURI := dp.sourceURI
+	length := dp.length
+	interval := dp.interval
+
+	go func() {
+		t := time.NewTicker(interval)
+		defer t.Stop()
+		for {
+			select {
+			case <-ctx.Done():
```

Comment:
we're counting on external elements to cancel this. we could introduce Done chan internally which would be closed on ReportComplete or ReportFailed

Reply:
Implemented your suggestion in 67f80f63f.  Let me know what you think.

Reply:
what about creating a subcontext that can be closed when the request is complete or cancelled (ofc we need to hold a ref to its cancellation function) ?

Reply:
> what about creating a subcontext that can be closed when the request is complete or cancelled (ofc we need to hold a ref to its cancellation function) ?

I like this pattern better as it doesn't require the consumer to remember to necessarily call either `ReportComplete` or `ReportFailed`.  Basically, it gives the consumer two options on how to cancel: either cancel the `ctx` passed to `Report` or call `ReportComplete`/`ReportFailed`. 

As far as implementation goes, we could do a subcontext or we use an internal `done` channel as implemented in https://github.com/elastic/elastic-agent/commit/67f80f63fe8c8359e381bc3233234c99ce4f1c41. Is one implementation necessarily better than the other?

Reply:
The context cancellation is more "abandon processing" and the done channel is the simple solution to unblock a go routine.

The context has the advantage of propagation through the subcontexts if the parent context is cancelled/closed/expires, the done channel has to be managed/closed explicitly.

Since you have already implemented the `done` channel you can keep that, I am just quite surprised to see that we only send an empty struct instead of closing the channel and removing the reference

Reply:
@pchila @michalpristas I updated the implementation such that there are now two ways to return from the goroutine inside `Report`:
* `Report` is passed a context.  When the context is done, the goroutine will return.
* When the consumer calls `ReportComplete` or `ReportFailed`, we close an internal `done` channel, which will also cause the goroutine in `Report` to return.